lib/google/gax/bundling.rb

```diff
@@ -0,0 +1,456 @@
+#     * Neither the name of Google Inc. nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+module Google
+  # Gax defines Google API extensions
+  module Gax
+    DEMUX_WARNING = [
+      'Warning: cannot demultiplex the bundled response, got ',
+      '%d subresponses; want %d, each bundled request will ',
+      'receive all responses'
+    ].join
+
+    # Helper function for #compute_bundle_id.
+    # Used to retrieve a nested field signified by name where dots in name
+    # indicate nested objects.
+    #
+    # @param obj [Object] an object.
+    # @param name [String] a name for a field in the object.
+    # @raise [NoMethodError] if the object does not contain the named attribute.
+    # @return [Object] value of named attribute. Can be nil.
+    def str_dotted_access(obj, name)
+      name.split('.').each do |part|
+        obj = obj.send(part)
+      end
+      obj.nil? ? nil : obj.to_s
+    end
+
+    # Computes a bundle id from the discriminator fields of `obj`.
+    #
+    # +discriminator_fields+ may include '.' as a separator, which is used to
+    # indicate object traversal.  This is meant to allow fields in the
+    # computed bundle_id.
+    # the return is an array computed by going through the discriminator fields
+    # in order and obtaining the str(value) object field (or nested object
+    # field) if any discriminator field cannot be found, ValueError is raised.
+    #
+    # @param obj [Object] an object.
+    # @param discriminator_fields [Array<String>] a list of discriminator
+    #     fields in the order to be to be used in the id.
+    # @raise [NoMethodError] if the object does not contain the named attribute.
+    # @return [Array<Object>] array of objects computed as described above.
+    def compute_bundle_id(obj, discriminator_fields)
+      result = []
+      discriminator_fields.each do |field|
+        result.push(str_dotted_access(obj, field))
+      end
+      result
+    end
+
+    # rubocop:disable Metrics/ClassLength
+
+    # Coordinates the execution of a single bundle.
+    #
+    # @attribute [r] bundle_id
```

Comment:
Strictly, `@!attribute` is the right format (see http://www.rubydoc.info/gems/yard/file/docs/Tags.md#attribute).